src/TraverseTracks.cpp

```diff
@@ -0,0 +1,153 @@
+#include "TraverseTracks.h"
+
+/**
+ * @brief Constructor for the TravseSegments class assigns the TrackGenerator
+ *        and pulls relevant information from it.
+ */
+TraverseTracks::TraverseTracks(TrackGenerator* track_generator) {
+
+  /* Save the track generator */
+  _track_generator = track_generator;
+
+  /* Determine the type of segment formation used */
+  _segment_formation = track_generator->getSegmentFormation();
+}
+
+
+/**
+ * @brief Destructor for TraverseTracks
+ */
+TraverseTracks::~TraverseTracks() {
+}
+
+
+/**
+ * @breif Loops over Tracks, applying the provided kernels to all segments and
+ *        the functionality described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme. If kernels are provided (not NULL) then they are deleted at
+ *          the end of the looping scheme.
+ * @param kernels MOCKernels to apply to all segments
+ */
+void TraverseTracks::loopOverTracks(MOCKernel** kernels) {
+
+  switch (_segment_formation) {
+    case EXPLICIT_2D:
+      loopOverTracks2D(kernels);
+      break;
+  }
+
+  if (kernels != NULL) {
+    delete kernels[0];
+    delete [] kernels;
+  }
+}
+
+
+/**
+ * @breif Loops over Tracks by parallel groups to avoid potential conflicts,
+ *        applying the provided kernels to all segments and the functionality
+ *        described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme. If kernels are provided (not NULL) then they are deleted at
+ *          the end of the looping scheme.
+ * @param kernels MOCKernels to apply to all segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup(MOCKernel** kernels) {
+  loopOverTracksByParallelGroup2D(kernels);
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks
+ * @details The onTrack(...) function is applied to all 2D Tracks and the
+ *          specified kernels are applied to all segments. If NULL is provided
+ *          for the kernels, only the onTrack(...) functionality is applied.
+ * @param kernels The MOCKernels dictating the functionality to apply to
+ *        segments
+ */
+void TraverseTracks::loopOverTracks2D(MOCKernel** kernels) {
+
+  /* Loop over all parallel tracks for each azimuthal angle */
+  Track** tracks_2D = _track_generator->getTracks();
+  int num_azim = _track_generator->getNumAzim();
+  for (int a=0; a < num_azim/2; a++) {
+    int num_xy = _track_generator->getNumX(a) + _track_generator->getNumY(a);
+#pragma omp for
+    for (int i=0; i < num_xy; i++) {
+
+      Track* track_2D = &tracks_2D[a][i];
+
+      /* Operate on segments if necessary */
+      if (kernels != NULL) {
+        kernels[0]->newTrack(track_2D);
```

Comment:
So this only uses the first kernel in the array? Why is the input an array of kernels then?